x-pack/plugins/infra/public/pages/metrics/hosts/hooks/use_hosts_url_state.ts

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useCallback, useEffect, useReducer } from 'react';
+import { TimeRange } from '@kbn/es-query';
+import DateMath from '@kbn/datemath';
+import deepEqual from 'fast-deep-equal';
+import * as rt from 'io-ts';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { fold } from 'fp-ts/lib/Either';
+import { constant, identity } from 'fp-ts/lib/function';
+import { enumeration } from '@kbn/securitysolution-io-ts-types';
+import { FilterStateStore } from '@kbn/es-query';
+import { useUrlState } from '../../../../utils/use_url_state';
+import { useKibanaTimefilterTime } from '../../../../hooks/use_kibana_timefilter_time';
+
+const DEFAULT_QUERY = {
+  language: 'kuery',
+  query: '',
+};
+const DEFAULT_FROM_MINUTES_VALUE = 15;
+const INITIAL_DATE = new Date();
+export const INITIAL_DATE_RANGE = { from: `now-${DEFAULT_FROM_MINUTES_VALUE}m`, to: 'now' };
+const CALCULATED_DATE_RANGE_FROM = new Date(
+  INITIAL_DATE.getMinutes() - DEFAULT_FROM_MINUTES_VALUE
+).getTime();
+const CALCULATED_DATE_RANGE_TO = INITIAL_DATE.getTime();
+
+const INITIAL_HOSTS_STATE: HostsState = {
+  query: DEFAULT_QUERY,
+  filters: [],
+  // for unified search
+  dateRange: { ...INITIAL_DATE_RANGE },
+  // for useSnapshot
+  dateRangeTimestamp: {
+    from: CALCULATED_DATE_RANGE_FROM,
+    to: CALCULATED_DATE_RANGE_TO,
+  },
+};
+
+type Action =
+  | {
+      type: 'setQuery';
+      payload: rt.TypeOf<typeof SetQueryType>;
+    }
+  | { type: 'setFilter'; payload: rt.TypeOf<typeof HostsFiltersRT> };
+
+const reducer = (state: HostsState, action: Action): HostsState => {
+  switch (action.type) {
+    case 'setFilter':
+      return { ...state, filters: [...action.payload] };
+    case 'setQuery':
+      const { filters, query, ...payload } = action.payload;
+      const newFilters = !filters ? state.filters : filters;
+      const newQuery = !query ? state.query : query;
+      return {
+        ...state,
+        ...payload,
+        filters: [...newFilters],
+        query: { ...newQuery },
+      };
+    default:
+      throw new Error();
+  }
+};
+
+export const useHostsUrlState = () => {
+  const [urlState, setUrlState] = useUrlState<HostsState>({
+    defaultState: INITIAL_HOSTS_STATE,
+    decodeUrlState,
+    encodeUrlState,
+    urlStateKey: '_a',
```

Comment:
I removed the comment here but this is still something we can discuss - should we use the `_a` to keep consistent with other apps state or we should use `hostFilters` for example similar to the metrics explorer approach 🤔

Reply:
I don't have a strong opinion about it. I like keeping consistency with other apps though. Wdyt @neptunian ?

Reply:
Doesn't matter to me, so long as we keep the current app names for those who have saved the url.